File: LeetCode/Others/476. Number Complement/476. Number Complement.java
class Solution {
    public int findComplement(int num) {
        
        String binary = convertToBi(num);
        
        StringBuilder sb = new StringBuilder();
        
        for (int i = 0; i < binary.length(); i++) {
            if (binary.charAt(i) == '0') {
                sb.append("0");
            } else if (binary.charAt(i) == '1') {
                sb.append("1");
            }
        }
        
        int result = Integer.parseInt(sb.toString(), 2);
        
        System.out.println(result);
        
        return result;
        
    }
    
    private String convertToBi(int num) {
        
        StringBuilder sb = new StringBuilder();
        
        while (num != 0) {
            int mod = num % 2;
            
            if (mod == 0) {
                sb.append("1");
            } else if (mod == 1) {
                sb.append("0");
            }
            
            num /= 2;
        }
        
        return sb.reverse().toString();
    }
}

File: LeetCode/Others/476. Number Complement/476. Number Complement.js
/**
 * @param {number} num
 * @return {number}
 */
var findComplement = function(num) {
  let binaryStr = Number(num).toString(2);
  let complementStr = '';
  for (let char of binaryStr) {
    if (char === '0') {
      complementStr += '1';
    } else {
      complementStr += '0';
    }
  }
  return parseInt(complementStr, 2);
};

/**
 * @param {number} num
 * @return {number}
 */
var findComplement = function(num) {
  let result = '';
  const binary = num.toString(2);
  for (let i = 0; i < binary.length; i++) {
    result += binary[i] ^ 1;
  }
  return parseInt(result, 2);
};

/**
 * @param {number} num
 * @return {number}
 */
var findComplement = function(num) {
  return parseInt(
    num
      .toString(2)
      .split('')
      .map(n => (n === '0' ? '1' : '0'))
      .join(''),
    2
  );
};

/**
 * @param {number} num
 * @return {number}
 * Bitwise and Math
 */
var findComplement = function(num) {
  let bit = 1;
  let result = 0;
  while (bit <= num) {
    if ((num & bit) === 0) {
      result += bit;
    }
    bit *= 2;
  }
  return result;
};

console.log(findComplement(5));
